cli/src/commands/git/sync.rs

```diff
@@ -0,0 +1,261 @@
+///
+/// This command fetches from Git remotes and rebases local commits that were
+/// descendants of remote-tracking bookmarks onto the new remote heads. This
+/// provides a workflow similar to `git pull --rebase` but operates on all
+/// tracked remote bookmarks simultaneously.
+///
+/// The rebase operation automatically drops any local commits that have been
+/// merged upstream.
+#[derive(clap::Args, Clone, Debug)]
+pub struct GitSyncArgs {
+    /// The remotes to sync with
+    ///
+    /// This defaults to the `git.fetch` setting. If that is not configured, and
+    /// if there are multiple remotes, the remote named "origin" will be used.
+    ///
+    /// By default, the specified remote names match exactly. Use a [string
+    /// pattern], e.g. `--remote 'glob:*'`, to select remotes using
+    /// patterns.
+    ///
+    /// [string pattern]:
+    ///     https://jj-vcs.github.io/jj/latest/revsets#string-patterns
+    #[arg(
+        long = "remote",
+        short = 'r',
+        value_name = "REMOTE",
+        value_parser = StringPattern::parse,
+        add = ArgValueCandidates::new(complete::git_remotes),
+    )]
+    remotes: Vec<StringPattern>,
+
+    /// Sync only these bookmarks, or bookmarks matching a pattern
+    ///
+    /// By default, the specified name matches exactly. Use `glob:` prefix to
+    /// expand `*` as a glob, e.g. `--branch 'glob:push-*'`. Other wildcard
+    /// characters such as `?` are *not* supported.
+    #[arg(
+        long = "bookmark",
+        short = 'b',
+        alias = "branch",
+        value_parser = StringPattern::parse,
+        add = ArgValueCandidates::new(complete::bookmarks),
+    )]
+    bookmarks: Vec<StringPattern>,
+
+    /// Sync with all remotes
+    #[arg(long, conflicts_with = "remotes")]
+    all_remotes: bool,
+}
+
+#[tracing::instrument(skip_all)]
+pub fn cmd_git_sync(
+    ui: &mut Ui,
+    command: &CommandHelper,
+    args: &GitSyncArgs,
+) -> Result<(), CommandError> {
+    let mut workspace_command = command.workspace_helper(ui)?;
+
+    // Determine which remotes to sync
+    let remote_patterns = if args.all_remotes {
+        vec![StringPattern::everything()]
+    } else if args.remotes.is_empty() {
+        get_default_fetch_remotes(ui, &workspace_command)?
+    } else {
+        args.remotes.clone()
+    };
+
+    let resolved_remotes =
+        resolve_remote_patterns(ui, workspace_command.repo().store(), &remote_patterns)?;
+    let remotes = resolved_remotes.iter().map(|r| r.as_ref()).collect_vec();
+
+    let mut tx = workspace_command.start_transaction();
+
+    // Capture the pre-fetch state of remote tracking bookmarks
+    let mut pre_fetch_heads: HashMap<RemoteRefSymbolBuf, CommitId> = HashMap::new();
+
+    for remote in &remotes {
+        for (name, remote_ref) in tx.repo().view().remote_bookmarks(remote) {
+            // Only capture non-conflicted targets
+            if let Some(commit_id) = remote_ref.target.as_normal() {
```

Comment:
I think we need to use the local bookmark tracking the remote. Remote bookmarks often point to hidden commits.